### PR TITLE
Init example for project template Node/Json/Vscode application

### PR DESCRIPTION
**Pull request by tortmayr**
Setup glsp server and client
Consume latest glsp server and client 
Load and save to a json resource
Supports creation and deletion, resizing, moving of nodes as well as label editing.
Part of eclipse-glsp/glsp#622

demo

Requires: https://github.com/eclipse-glsp/glsp-vscode-integration/pull/26
Requires: https://github.com/eclipse-glsp/glsp-server-node/pull/16
Contributed on behalf of STMicroelectronics.

### CODE REVIEW

File: project-templates/node-vscode-json/README.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    Similar to the `Launch Tasklist Diagram Extension` but does not start the GLSP server as embedded process.
+    It expects that the GLSP Server process is already running and has been started externally with the `Launch Tasklist GLSP Server` config.
+- `Launch Tasklist GLSP Server`<br>
+    This config can be used to manually launch the `Tasklist GLSP Server` node process.
+    Breakpoints the source files of the `glsp-server` directory will be picked up.
+    In order to use this config, the `Tasklist Diagram Extension` has to be launched in `External` server mode.
+- `Launch Tasklist Diagram extension with external GLSP Server`<br>
+    This is a convienience compound config that launches both the `Tasklist Diagram Extension` in external server mode and the
+    `Tasklist GLSP server` process.
+    Enables debugging of both the `glsp-client` and `glsp-server`code simultaneously.
```
Comment (planger) [inline project-templates/node-vscode-json/README.md:86]: I guess this is not yet up to date, is it?

File: workflow/glsp-client/workflow-glsp/src/test-ui-extension/di.config.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
+/********************************************************************************
```
Comment (planger) [inline workflow/glsp-client/workflow-glsp/src/test-ui-extension/di.config.ts:1]: I guess this module was committed by accident?

File: project-templates/node-vscode-json/glsp-client/tasklist-vscode/extension/package.json [hunk 3]
```diff
@@ -0,0 +1,162 @@
+{
```
Comment (planger) [inline project-templates/node-vscode-json/glsp-client/tasklist-vscode/extension/package.json:1]: I think the webview.js (above, can't add a comment there directly) shouldn't be committed right? Isn't it generated by the build?

**Review by planger — COMMENTED** [review]
I didn't have the chance to test it yet, but looks great so far! Thanks!

File: project-templates/node-vscode-json/README.md (hunk 1 above)
Comment (tortmayr) [reply project-templates/node-vscode-json/README.md]: Pushed an updated version. The only thing missing now is the intro section and the screencast

File: workflow/glsp-client/workflow-glsp/src/test-ui-extension/di.config.ts (hunk 2 above)
Comment (tortmayr) [reply workflow/glsp-client/workflow-glsp/src/test-ui-extension/di.config.ts]: Right. Is removed now

File: project-templates/node-vscode-json/glsp-client/tasklist-vscode/extension/package.json (hunk 3 above)
Comment (tortmayr) [reply project-templates/node-vscode-json/glsp-client/tasklist-vscode/extension/package.json:1]: Good catch.  Has been removed and added to .gitignore

**Comment by tortmayr** [comment]
Thanks for the initial review. I pushed an update that also consumes the required PRs. You should now be able to build and run the example.

File: project-templates/node-vscode-json/glsp-server/src/diagram/tasklist-diagram-module.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+export class TasklistDiagramModule extends DiagramModule {
+    readonly diagramType = 'minimal-diagram';
+
+    protected override configure(
+        bind: interfaces.Bind,
+        unbind: interfaces.Unbind,
+        isBound: interfaces.IsBound,
+        rebind: interfaces.Rebind
+    ): void {
+        bind(TasklistModelIndex).toSelf().inSingletonScope();
```
Comment (planger) [inline project-templates/node-vscode-json/glsp-server/src/diagram/tasklist-diagram-module.ts:52]: Why is this binding before the super-call? This looks surprising. Is this required to ensure that when `bindGModelIndex()` is called, this already exists?

This isn't a blocking issue, but I think it'd be much nicer if that all could be handled in `bindGModelIndex()`.
For instance, wouldn't it work if all those methods below would get `bind`, `unbind`, etc. provided as parameter? This way, they could at least bind their services in the same method before they return the class to be bound as service?

File: project-templates/node-vscode-json/glsp-server/src/diagram/tasklist-diagram-module.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        return TasklistDiagramConfiguration;
+    }
+
+    protected bindSourceModelStorage(): BindingTarget<SourceModelStorage> {
+        return TasklistStorage;
+    }
+    protected bindModelState(): BindingTarget<ModelState> {
+        return TasklistModelState;
+    }
+    protected bindGModelFactory(): BindingTarget<GModelFactory> {
```
Comment (planger) [inline project-templates/node-vscode-json/glsp-server/src/diagram/tasklist-diagram-module.ts:79]:
```suggestion

        protected bindGModelFactory(): BindingTarget<GModelFactory> {
```

File: project-templates/node-vscode-json/glsp-server/src/handler/create-task-node-handler.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import * as uuid from 'uuid';
+import { Task } from '../model/tasklist-model';
+import { TasklistModelState } from '../model/tasklist-model-state';
+
+@injectable()
+export class CreateTaskHandler extends CreateNodeOperationHandler {
+    @inject(TasklistModelState)
+    protected override modelState: TasklistModelState;
+
+    readonly elementTypeIds = [DefaultTypes.NODE];
```
Comment (planger) [inline project-templates/node-vscode-json/glsp-server/src/handler/create-task-node-handler.ts:27]: I'd put this as a first member of the class

**Review by planger — CHANGES_REQUESTED** [review]
Looks great, but unfortunately I get a build error (below)!
Just a few minors inline and the following things:
  * Root `yarn.lock` is missing.
  * glsp-client/tasklist-vscode/extension/pack/webview.js is still there
  * `workflow-diagram-widget.ts` et al is still there

```
~/Git/OpenSource/eclipse-glsp/glsp-examples/project-templates/node-vscode-json (node-vscode ✔) git clean -xfd
~/Git/OpenSource/eclipse-glsp/glsp-examples/project-templates/node-vscode-json (node-vscode ✔) yarn build
yarn run v1.22.18
warning package.json: No license field
$  yarn build:client && yarn build:server
warning package.json: No license field
$ yarn --cwd glsp-client install
[1/5] Validating package.json...
[2/5] Resolving packages...
warning Resolution field "sprotty@0.12.0-next.00a6590.0" is incompatible with requested version "sprotty@~0.11.1"
[3/5] Fetching packages...
warning Pattern ["rimraf@latest"] is trying to unpack in the same destination "/home/philip/.cache/yarn/v6/npm-rimraf-3.0.2-f1a5402ba6220ad52cc1282bac1ae3aa49fd061a-integrity/node_modules/rimraf" as pattern ["rimraf@^3.0.2","rimraf@^3.0.2","rimraf@^3.0.2","rimraf@^3.0.2","rimraf@^3.0.2","rimraf@^3.0.2"]. This could result in non-deterministic behavior, skipping.
warning Pattern ["sprotty-protocol@next"] is trying to unpack in the same destination "/home/philip/.cache/yarn/v6/npm-sprotty-protocol-0.12.0-next.00a6590.0-31d65d864a8c662f0621ed313b8315a72ca97689-integrity/node_modules/sprotty-protocol" as pattern ["sprotty-protocol@0.12.0-next.00a6590.0"]. This could result in non-deterministic behavior, skipping.
warning Pattern ["@eclipse-glsp/protocol@next"] is trying to unpack in the same destination "/home/philip/.cache/yarn/v6/npm-@eclipse-glsp-protocol-0.10.0-next.b3a810f.172-ee4ee6d4211e0bbd61000782f1d5fbea2ce5298c-integrity/node_modules/@eclipse-glsp/protocol" as pattern ["@eclipse-glsp/protocol@0.10.0-next.b3a810f.172+b3a810f","@eclipse-glsp/protocol@0.10.0-next.b3a810f.172+b3a810f"]. This could result in non-deterministic behavior, skipping.
warning @eclipse-glsp/tasklist-vscode@0.9.0: The engine "vscode" appears to be invalid.
warning @eclipse-glsp/vscode-integration@0.10.0-next.4484750.44: The engine "vscode" appears to be invalid.
[4/5] Linking dependencies...
warning "@eclipse-glsp/config > @eclipse-glsp/mocha-config@0.10.0-next.b7bb985.103" has unmet peer dependency "@types/node@>=12.x".
warning "@eclipse-glsp/config > @eclipse-glsp/mocha-config@0.10.0-next.b7bb985.103" has unmet peer dependency "reflect-metadata@>=0.1.13".
warning "@eclipse-glsp/config > @eclipse-glsp/mocha-config > ts-node@10.7.0" has unmet peer dependency "@types/node@*".
[5/5] Building fresh packages...
$ lerna run prepare
lerna notice cli v4.0.0
lerna info Executing command in 3 packages: "yarn run prepare"
@eclipse-glsp-examples/tasklist-glsp: $ yarn clean && yarn build && yarn lint
@eclipse-glsp-examples/tasklist-glsp: $ rimraf lib tsconfig.tsbuildinfo
@eclipse-glsp-examples/tasklist-glsp: $ tsc
@eclipse-glsp-examples/tasklist-glsp: $ eslint -c .eslintrc.js --ext .ts,.tsx ./src
@eclipse-glsp/tasklist-webview: $ yarn clean && yarn build && yarn lint
@eclipse-glsp/tasklist-webview: $ rimraf lib pack
@eclipse-glsp/tasklist-webview: $ tsc && webpack --mode=development
@eclipse-glsp/tasklist-webview: error TS2688: Cannot find type definition file for 'node'.
@eclipse-glsp/tasklist-webview:   The file is in the program because:
@eclipse-glsp/tasklist-webview:     Entry point of type library 'node' specified in compilerOptions
@eclipse-glsp/tasklist-webview: error Command failed with exit code 2.
@eclipse-glsp/tasklist-webview: info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
@eclipse-glsp/tasklist-webview: error Command failed with exit code 2.
@eclipse-glsp/tasklist-webview: info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
lerna ERR! yarn run prepare exited 2 in '@eclipse-glsp/tasklist-webview'
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
~/Git/OpenSource/eclipse-glsp/glsp-examples/project-templates/node-vscode-json (node-vscode ✔) 
```

**Comment by planger** [comment]
I've pushed a change that fixes the build for me (d53fe07).
Looks good now from a functionality point of view. However, I get an error when I add a node and then try to edit a label. Seems the index isn't updated after operations?
```
6:43:33 PM error -  [DefaultGLSPServer] Could not process action: 'requestEditValidation Could not retrieve element with id: '16588fa3-1b59-4f95-9309-03a7409d3860'
            Error: Could not retrieve element with id: '16588fa3-1b59-4f95-9309-03a7409d3860'
    at getOrThrow (/home/philip/Git/OpenSource/eclipse-glsp/glsp-examples/project-templates/node-vscode-json/glsp-server/node_modules/@eclipse-glsp/server-node/lib/utils/glsp-server-error.js:30:11)
    at TasklistModelIndex.get (/home/philip/Git/OpenSource/eclipse-glsp/glsp-examples/project-templates/node-vscode-json/glsp-server/node_modules/@eclipse-glsp/server-node/lib/features/model/gmodel-index.js:86:51)
    at ValidateLabelEditAdapter.validate (/home/philip/Git/OpenSource/eclipse-glsp/glsp-examples/project-templates/node-vscode-json/glsp-server/node_modules/@eclipse-glsp/server-node/lib/features/directediting/context-edit-validator.js:32:47)
    at RequestEditValidationHandler.execute (/home/philip/Git/OpenSource/eclipse-glsp/glsp-examples/project-templates/node-vscode-json/glsp-server/node_modules/@eclipse-glsp/server-node/lib/features/directediting/request-edit-validation-handler.js:39:48)
    at DefaultActionDispatcher.executeHandler (/home/philip/Git/OpenSource/eclipse-glsp/glsp-examples/project-templates/node-vscode-json/glsp-server/node_modules/@eclipse-glsp/server-node/lib/actions/action-dispatcher.js:68:47)
    at DefaultActionDispatcher.doDispatch (/home/philip/Git/OpenSource/eclipse-glsp/glsp-examples/project-templates/node-vscode-json/glsp-server/node_modules/@eclipse-glsp/server-node/lib/actions/action-dispatcher.js:58:41)
    at Object.promise (/home/philip/Git/OpenSource/eclipse-glsp/glsp-examples/project-templates/node-vscode-json/glsp-server/node_modules/@eclipse-glsp/server-node/lib/actions/action-dispatcher.js:48:52)
    at PromiseQueue.resolveQueuedPromise (/home/philip/Git/OpenSource/eclipse-glsp/glsp-examples/project-templates/node-vscode-json/glsp-server/node_modules/@eclipse-glsp/server-node/lib/utils/promise-queue.js:68:41)
    at PromiseQueue.dequeue (/home/philip/Git/OpenSource/eclipse-glsp/glsp-examples/project-templates/node-vscode-json/glsp-server/node_modules/@eclipse-glsp/server-node/lib/utils/promise-queue.js:57:18)
    at /home/philip/Git/OpenSource/eclipse-glsp/glsp-examples/project-templates/node-vscode-json/glsp-server/node_modules/@eclipse-glsp/server-node/lib/utils/promise-queue.js:44:18
 ```

**Comment by planger** [comment]
One last thing that I've observed: I think `tsconfig.tsbuildinfo` in glsp-server is not gitignored.

**Comment by planger** [comment]
Could it be that animation is broken? Maybe because we get new IDs after each change?

File: project-templates/node-vscode-json/glsp-server/src/diagram/tasklist-diagram-module.ts (hunk 4 above)
Comment (tortmayr) [reply project-templates/node-vscode-json/glsp-server/src/diagram/tasklist-diagram-module.ts:52]: I aggree that is workaround is not nice and I already have openend a GLSP issue for that (https://github.com/eclipse-glsp/glsp/issues/631). 
For now we can get around this by using the binding context that is provided by the super implementation. I will update this accordingly

**Comment by tortmayr** [comment]
Thanks for the fast review Philip. I have pushed an update that should address the remaining issues

File: project-templates/node-vscode-json/glsp-server/tsconfig.tsbuildinfo [hunk 7]
```diff
@@ -0,0 +1 @@
...9,170],[293,171],[286,172],[287,173],[285,174],[294,167]],"semanticDiagnosticsPerFile":[154,155,156,159,161,163,162,164,165,166,160,157,153,167,168,169,170,158,171,52,53,56,57,58,59,61,62,63,64,65,66,74,67,68,69,70,60,71,54,55,72,73,75,152,148,147,149,150,151,51,199,198,193,201,202,204,203,216,213,217,243,215,214,250,251,218,252,253,223,229,222,224,254,228,232,231,230,255,233,256,257,234,219,220,221,235,258,236,259,238,237,239,240,260,261,241,262,242,263,264,265,266,267,268,269,270,271,273,272,283,212,274,275,276,277,226,278,227,225,200,245,249,246,196,247,248,244,279,194,280,211,281,195,197,282,210,300,77,78,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,79,126,95,96,97,127,98,99,100,101,102,103,104,105,106,107,108,109,111,110,112,113,114,115,116,117,118,119,120,121,122,123,124,125,284,188,180,177,183,179,181,185,178,184,182,175,172,174,176,173,192,186,189,191,187,190,205,298,47,43,48,46,50,41,44,40,42,45,49,8,10,9,2,11,12,13,14,15,16,17,18,3,4,22,19,20,21,23,24,25,5,26,27,28,29,6,30,31,32,33,7,38,34,35,36,37,1,39,132,129,133,136,130,131,128,134,135,76,145,137,146,144,138,142,143,140,139,141,206,209,207,208,297,295,296,288,289,290,291,292,299,293,286,287,285,294]},"version":"4.6.3"}
```
[the remark this answers is not in the document]
Comment (planger) [reply project-templates/node-vscode-json/glsp-server/tsconfig.tsbuildinfo]: Fixed with [11df608](https://github.com/eclipse-glsp/glsp-examples/pull/110/commits/11df6085183b80da9ee97e7ee09ce17b5b7caa57)

File: workflow/glsp-client/workflow-theia/src/browser/diagram/workflow-task-editing-context-menu.ts [hunk 8]
```diff
@@ -32,7 +32,10 @@ export class WorkflowTaskEditCommandContribution implements CommandContribution
         commands.registerCommand(
```
[the remark this answers is not in the document]
Comment (planger) [reply workflow/glsp-client/workflow-theia/src/browser/diagram/workflow-task-editing-context-menu.ts]: Fixed with [11df608](https://github.com/eclipse-glsp/glsp-examples/pull/110/commits/11df6085183b80da9ee97e7ee09ce17b5b7caa57)

File: workflow/glsp-client/workflow-glsp/src/index.ts [hunk 9]
```diff
@@ -15,7 +15,7 @@
  ********************************************************************************/
```
[the remark this answers is not in the document]
Comment (planger) [reply workflow/glsp-client/workflow-glsp/src/index.ts]: Fixed with [11df608](https://github.com/eclipse-glsp/glsp-examples/pull/110/commits/11df6085183b80da9ee97e7ee09ce17b5b7caa57)

File: workflow/glsp-client/workflow-theia/src/browser/diagram/workflow-diagram-widget.ts [hunk 10]
```diff
@@ -0,0 +1,26 @@
+/********************************************************************************
```
[the remark this answers is not in the document]
Comment (planger) [reply workflow/glsp-client/workflow-theia/src/browser/diagram/workflow-diagram-widget.ts]: Fixed with [11df608](https://github.com/eclipse-glsp/glsp-examples/pull/110/commits/11df6085183b80da9ee97e7ee09ce17b5b7caa57)

**Comment by tortmayr** [comment]
Thanks Philip. I have pushed an update that fixes the update animations

**Pull request merged**